File: app/src/main/java/com/example/ruslanyussupov/popularmovies/di/DataModule.java
package com.example.ruslanyussupov.popularmovies.di;

import android.arch.persistence.room.Room;
import android.content.Context;

import com.example.ruslanyussupov.popularmovies.data.DataSource;
import com.example.ruslanyussupov.popularmovies.data.Repository;
import com.example.ruslanyussupov.popularmovies.data.local.FavouriteMoviesDb;
import com.example.ruslanyussupov.popularmovies.data.remote.TheMovieDbService;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import retrofit2.Retrofit;

@Module
public class DataModule {

    @Provides
    @Singleton
    public TheMovieDbService provideTheMovieDbService(Retrofit retrofit) {
        return retrofit.create(TheMovieDbService.class);
    }

    @Provides
    @Singleton
    public DataSource provideDataSource() {
        return new Repository();
    }

    @Provides
    @Singleton
    public FavouriteMoviesDb provideFavouriteMovieDb(Context appContext) {
        return Room.databaseBuilder(appContext, FavouriteMoviesDb.class,
                "fav_movies").build();
    }

}

File: app/src/main/java/com/example/ruslanyussupov/popularmovies/di/DataModule.kt
package com.example.ruslanyussupov.popularmovies.di

import androidx.room.Room
import android.content.Context
import android.content.SharedPreferences

import com.example.ruslanyussupov.popularmovies.data.DataSource
import com.example.ruslanyussupov.popularmovies.data.Repository
import com.example.ruslanyussupov.popularmovies.data.local.MovieDb
import com.example.ruslanyussupov.popularmovies.data.remote.TheMovieDbService

import javax.inject.Singleton

import dagger.Module
import dagger.Provides
import retrofit2.Retrofit

@Module
class DataModule {

    @Provides
    @Singleton
    fun provideTheMovieDbService(retrofit: Retrofit): TheMovieDbService {
        return retrofit.create(TheMovieDbService::class.java)
    }

    @Provides
    @Singleton
    fun provideDataSource(): DataSource {
        return Repository()
    }

    @Provides
    @Singleton
    fun provideMovieDb(appContext: Context): MovieDb {
        return Room.databaseBuilder(appContext, MovieDb::class.java,
                "movies").build()
    }

    @Provides
    @Singleton
    fun provideRequestSharedPrefs(appContext: Context): SharedPreferences {
        return appContext.getSharedPreferences("request_prefs", Context.MODE_PRIVATE)
    }

}
